types/csmim.obj.registration.1.yaml

```diff
@@ -0,0 +1,86 @@
+%YAML 1.2
+---
+id: csmim.obj.registration.1
+supertypes: []
+description: CSMIM registration service interface as specified in ARINC 853
+attributes: []
+resources:
+  - id: claim
+    mode: x
+    type: void
```

Comment:
if we would like to return feedback for failed or partial successful registrations, could we add an option here?

Reply:
You mean the error return payload? That is defined by the standard, but it is not captured in the "type" property. So, according to the standard, if the type is "void" that means the successful response payload is empty. But the error response payload is always either empty or a dict, regardless of the value of the "type" property.

Do you think it would be good to describe the error response payload in the object type, too? For example using a "error-value" or "error-values" entry?

Reply:
By the way, as you can see I wanted to include a detailed description of the parameters using "dict-values". But that does not pass our validation script. So I had to comment that out. If we agreed on using something like "dict-values" or "dict-items" or "dict-entries" and patched the validation script accordingly, that would allow us to be more precise here. Similar for "error-values" or "error-items" or "error-entries". See also issue #6.

Reply:
> You mean the error return payload? That is defined by the standard, but it is not captured in the "type" property. So, according to the standard, if the type is "void" that means the successful response payload is empty. But the error response payload is always either empty or a dict, regardless of the value of the "type" property.
> 
> Do you think it would be good to describe the error response payload in the object type, too? For example using a "error-value" or "error-values" entry?

I have similar question as Maththias. It's not just for "error condition", but including successful registrations. ARINC853 Table 6-4 definition for 'void' states "Resource has no value respectively returns no value when executed". However in case of full registration or registration renewal, the operation will optionally return a "code" as a user property (refer Table 6-13, 6-15, 6-17).  How do we capture that in the type defintion? 

On a related note, I am wondering if we should have another separate type definition for "registration response" as outlined in ARINC 853 Table 6-16 – Payload structure of an error REGISTER response message?

Reply:
Hmm well I think this is suboptimal wording in the standard. The reponse "code" is mandatory, and it has nothing to do with a returned value. The value is in the payload. Resource type = void means execute response payload = empty. The response code must always be supplied. If it is 2xx: success response (currently we use only 200). Otherwise: error response. If it is an error response, the payload is not determined by the resource type, but must be either empty or a dictionary that should include a human-readable "error" message and may include additional information. The additional information can currently only be described in the resource type's "description".

Reply:
This is true for registration responses and for any other response. Always empty or a dict. We do not need another type. But if the error response contains additional information, it could be useful to describe that additional information more formally in the type definition. For the same reasons that it could be useful to describe other dict payloads more formally in the type defintion.

Reply:
my comment was addressing the case where a user requests multiple objects and one of them is rejected. In this case, a response could be partially successful and the return payload could inform about the failed part. Alternatively, the complete registration must fail in order to return this information.

Reply:
Ah okay. Current definition is that the whole registration must fail if one path is rejected (conflict or unauthorized). The client gets the information which path is the problem, and can retry. Until all conflicting parts have been removed.

If we wanted to have partially successful registrations:

- that would be a change to the standard.
- the resource type must be dict.

If we wanted to have the option of either returning nothing or a dict, that would be another change to the standard. In this case, to save one byte (an empty CBOR dict).

Reply:
OK. Let's keep it as proposed, then.